vllm/worker/worker.py

```diff
@@ -238,22 +238,44 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
         # Get the peak memory allocation recorded by torch
         peak_memory = torch.cuda.memory_stats()["allocated_bytes.all.peak"]
 
+        # Edge case: Check for any memory left around that may have been
+        # allocated on the gpu outside of `torch`
+        torch.cuda.empty_cache()
+        leftover_allocations = torch.cuda.mem_get_info(
+        )[0] - self.init_gpu_memory
+        if leftover_allocations > 0:
+            logger.info(
+                "Found %.2f GB of allocated memory leftover after clearing "
+                "torch cache. Adding to peak memory usage.",
+                leftover_allocations / (1024**3))
+            peak_memory += leftover_allocations
+
+        available_kv_cache_memory = (
+            total_gpu_memory * self.cache_config.gpu_memory_utilization -
+            peak_memory)
+
+        logger.info("Initial memory usage before profile: %.2f GB",
+                    (total_gpu_memory - self.init_gpu_memory) / (1024**3))
+        logger.info("Peak memory usage during profile: %.2f GB",
+                    peak_memory / (1024**3))
+        logger.info(
+            "Available memory for KV cache with %.2f gpu utilization: %.2f GB",
+            self.cache_config.gpu_memory_utilization,
+            available_kv_cache_memory / (1024**3))
+
         cache_block_size = self.get_cache_block_size_bytes()
         if cache_block_size == 0:
             num_gpu_blocks = 0
             num_cpu_blocks = 0
         else:
-            num_gpu_blocks = int(
-                (total_gpu_memory * self.cache_config.gpu_memory_utilization -
-                 peak_memory) // cache_block_size)
+            num_gpu_blocks = int(available_kv_cache_memory // cache_block_size)
             num_cpu_blocks = int(self.cache_config.swap_space_bytes //
                                  cache_block_size)
         num_gpu_blocks = max(num_gpu_blocks, 0)
         num_cpu_blocks = max(num_cpu_blocks, 0)
         if self.model_runner.lora_manager:
             self.model_runner.remove_all_loras()
         gc.collect()
```

Comment:
Probably this `gc.collect()` should also be moved up above `torch.cuda.empty_cache()`?

Also I'm not sure what the reason for the `remove_all_loras()` is here.. perhaps that should also be moved up to before the gc is done?

Reply:
yeah I'm not sure on that either, I'm guessing we're just trying to clean up everything we did during profiling, before the KV cache is allocated?

re: moving the gc.collect(), I was trying to leave it later here in case there was something allocated outside torch that hadn't been GC'ed yet that we may need to account for in the peak memory usage. If we run all the cleanup and then check the free memory, then the only reason it would be lower is if there's a memory leak, right?

idk- I could go either way. I'm not 100% sold we need the extra check for non-torch allocated memory since it's pretty flaky to try to check for. Think we should just back that out and leave the `torch.empty_cache()` down here?

Reply:
update: @tjohnson31415 will give this a go to see if he can reproduce the NCCL allocations he was seeing that were blowing up vram usage. If this code catches it we'll keep it in, if not I'll back it out

Reply:
It needs [a small fix](https://github.com/vllm-project/vllm/pull/9352/files#r1803590168), but the `leftover_allocations` does pick up on the extra memory allocated by NCCL during the profile run. When it is not accounted for, I got an OOM when allocating the KV Cache...

The call to `remove_all_loras()` seems like it would make more sense to have in `profile_run()`. Any tensors allocated for LoRA would be included in the torch allocated peak.

In my test with Llama-3.1-8B-Instruct w/ TP=8, moving `gc.collect()` and `remove_all_loras()` above `leftover_allocations` made no difference in the printed messages, so they only clean up a small amount of memory if anything.

Reply:
🌶️🌶️🌶️!

> The call to remove_all_loras() seems like it would make more sense to have in profile_run(). Any tensors allocated for LoRA would be included in the torch allocated peak.

I agree the cleanup in general could better live in the profile run executions, but I do want to limit the blast radius here to this file. I'll leave as-is unless anybody feels strongly about refactoring into the individual model runners